### PR TITLE
swat/feature/typescript

### PR DESCRIPTION
**Pull request by sshankar19**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,30 +1,38 @@
...
+        "reflect-metadata": "^0.1.12"
     },
     "devDependencies": {
+        "@types/chai": "^3.4.34",
+        "@types/mocha": "^2.2.39",
+        "@types/node": "^7.0.5",
         "chai": "*",
         "eslint": "3.7.x",
         "istanbul": "0.4.5",
+        "q": "1.x",
```
Comment (hackerrdave) [inline package.json]: why is "q" being added?

File: package.json [hunk 2]
```diff
@@ -1,30 +1,38 @@
...
+        "sinon": "1.15.x",
+        "typescript": "^2.8.4",
```
Comment (hackerrdave) [inline package.json]: should lock to specific version of typescript (same with all modules)

File: tsconfig.json [hunk 3]
```diff
@@ -1,16 +1,24 @@
...
-    "baseUrl": ".",
+    "lib": ["es2015", "es2017", "dom"],
     "noImplicitAny": false,
-    "sourceMap": true,
     "typeRoots": [
       "../node_modules/@types"
     ],
+    "declaration": true,
+    "declarationDir": "dist",
+    "outDir": "./dist",
```
Comment (hackerrdave) [inline tsconfig.json]: we'll need to add this dir to .gitignore correct?

File: test/typescript/model/Ark.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import {Supertype, supertypeClass, property} from '../../../index';
+import {Supertype, supertypeClass, property} from '../../../dist/index';
```
Comment (hackerrdave) [inline test/typescript/model/Ark.ts]: why importing from "dist" directory?

File: test/typescript/animals.ts [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
-import {LionContainer} from "./model/Arc";
-import {AnimalContainer} from "./model/Arc";
+import {Ark} from "./model/Ark";
+import {LionContainer} from "./model/Ark";
+import {AnimalContainer} from "./model/Ark";
 import {Lion} from "./model/Lion";
 import {Bear} from "./model/Bear";
 import {Animal} from "./model/Animal";
-import {amorphicStatic} from "../../index";
+import {amorphicStatic} from "../../dist/index";
```
Comment (hackerrdave) [inline test/typescript/animals.ts]: why from "dist" dir?

File: test/animals.js [hunk 6]
```diff
@@ -1,5 +1,5 @@
 var expect = require('chai').expect;
-var ObjectTemplate = require('../index.js');
+var ObjectTemplate = require('../dist/index.js').default;
```
Comment (hackerrdave) [inline test/animals.js]: should we take this time to convert all tests to typescript as well?

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * ObjectTemplate - n Type System with Benefits
+ */
+
+import {ObjectTemplate} from './ObjectTemplate';
+
+ObjectTemplate.init();
+export default ObjectTemplate;
+
+let amorphicStatic = ObjectTemplate.amorphicStatic;
+let SupertypeSession = ObjectTemplate.amorphicStatic;
```
Comment (hackerrdave) [inline src/index.ts]: do you know what this is used for?

File: src/decorators.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+            getDictionary();
+            return target.__shadowParent__;
+        }
+        function getChildren() {
+            getDictionary();
+            return target.__shadowChildren__;
+        }
+
+        /*
+        TODO: Typescript
```
Comment (hackerrdave) [inline src/decorators.ts]: we should track this TODO in ticket

File: package.json (hunk 1 above)
Comment (sshankar19) [reply package.json]: Q should be removed, will remove

File: package.json (hunk 2 above)
Comment (sshankar19) [reply package.json]: locking it to 2.9.2

File: tsconfig.json (hunk 3 above)
Comment (sshankar19) [reply tsconfig.json]: already added

File: test/typescript/model/Ark.ts (hunk 4 above)
Comment (sshankar19) [reply test/typescript/model/Ark.ts]: This is for compilation purposes for the tests only.  The dist/index is the entry point of our application (and what gets published) so it makes sense for our tests for supertype to import and use supertype the same way our modules do.

File: src/index.ts (hunk 7 above)
Comment (sshankar19) [reply src/index.ts]: SupertypeSession is used or at least referenced in the other amorphic components (amorphic and persistor). I need to export this for Remoteable.ts and Persistable.ts. I might clean that up later when I clean up Amorphic and Persistor, but at the moment need to export it.

File: src/decorators.ts (hunk 8 above)
Comment (sshankar19) [reply src/decorators.ts]: Right, I'm not too sure what this TODO is for anymore. it's been too long.

File: test/typescript/animals.ts (hunk 5 above)
Comment (sshankar19) [reply test/typescript/animals.ts]: answered in a different comment

File: test/animals.js (hunk 6 above)
Comment (sshankar19) [reply test/animals.js]: We have some tests in typescript, I added them to circle ci config (they all pass).  Let's keep these tests in js. I don't want to convert them at the moment either, that's not in the scope of this mr

Comment (hackerrdave) [reply test/animals.js]: ok sounds good

File: src/serializer.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+    for (var propb in pojo) {
+        propb = propb.replace(/^__/, '');
+        var defineProp = props[propb];
+        if (!defineProp)
+            continue;
+        var type = defineProp.type;
+
+        // Because semotus can serialize only the shadow properties we try and restore them
+        var pojoProp = (type && typeof pojo['__' + propb] !== 'undefined') ? '__' + propb : propb;
```
Comment (srksag) [inline src/serializer.ts]: May be a "Todo" item for future, if semotus needs these shadow properties to be restored, the functionality should be moved to semotus. We should clearly segregate the responsibilities of these libraries.

File: src/Supertype.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+function constructorName(constructor) {
+    var namedFunction = constructor.toString().match(/function ([^(]*)/);
+    return namedFunction ? namedFunction[1] : null;
+}
+
+export type Constructable = new (...args: any[]) => {};
+
+
+/**
+ * This is the base class for typescript classes.  It must
```
Comment (nsaenz7) [inline src/Supertype.ts]: was there supposed to be more to this comment that got truncated accidentally?

Comment (sshankar19) [reply src/Supertype.ts]: hahaha yep that's what it looks like.

Comment (sshankar19) [reply src/Supertype.ts]: I'll just delete the it must part

**Pull request merged**